src/System.Text.Json/src/System/Text/Json/Node/JsonBoolean.cs

```diff
@@ -0,0 +1,81 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json
+{
+    /// <summary>
+    ///   Represents a boolean JSON value.
+    /// </summary>
+    public class JsonBoolean : JsonNode, IEquatable<JsonBoolean>
+    {
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonBoolean"/> class representing the value <see langword="false"/>.
+        /// </summary>
+        public JsonBoolean() => Value = false;
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="JsonBoolean"/> class representing a specified value.
+        /// </summary>
+        public JsonBoolean(bool value) => Value = value;
+
+        /// <summary>
+        ///   Gets or sets the boolean value represented by the instance.
+        /// </summary>
+        public bool Value { get; set; }
+
+        /// <summary>
+        ///   Converts a <see cref="bool"/> to a JSON boolean.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonBoolean(bool value) => new JsonBoolean(value);
+
+        /// <summary>
+        ///   Compares <paramref name="obj"/> to the value of this instance. 
+        /// </summary>
+        /// <param name="obj">The object to compare against.</param>
+        /// <returns>
+        ///   <see langword="true"/> if the boolean value of this instance matches <paramref name="obj"/>,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public override bool Equals(object obj) => obj is JsonBoolean jsonBoolean && Value == jsonBoolean.Value;
+
+        /// <summary>
+        ///   Calculates a hash code of this instance.
+        /// </summary>
+        /// <returns>A hash code for this instance.</returns>
+        public override int GetHashCode() => Value.GetHashCode();
+
+        /// <summary>
+        ///   Compares other JSON boolean to the value of this instance. 
+        /// </summary>
+        /// <param name="other">The JSON boolean to compare against.</param>
+        /// <returns>
+        ///   <see langword="true"/> if the boolean value of this instance matches <paramref name="other"/>,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public bool Equals(JsonBoolean other) => !(other is null) && Value == other.Value;
+
+        /// <summary>
+        ///   Compares values of two JSON booleans. 
+        /// </summary>
+        /// <param name="left">The JSON boolean to compare.</param>
+        /// <param name="right">The JSON boolean to compare.</param>
+        /// <returns>
+        ///   <see langword="true"/> if values of instances match,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public static bool operator ==(JsonBoolean left, JsonBoolean right) => left?.Value == right?.Value;
```

Comment:
Ok, thanks, I will apply it to `JsonNumber` as well.

Reply:
Should I add annotation about inlining as well?

Reply:
I am not sure if it is necessary. Looks like it doesn't help, so not adding it is fine.

@benaadams - you added that attribute to SortVersion here, along with a bunch of other places (https://github.com/dotnet/coreclr/pull/21765), was it necessary?

```C#
public class Foo
{
    int _field;

    public Foo(int field)
    {
        _field = field;
    }

    public override bool Equals(object obj)
        => obj is Foo otherVersion && Equals(otherVersion);

    [MethodImpl(MethodImplOptions.NoInlining)]
    public bool Equals(Foo other)
    {
        if (other == null)
        {
            return false;
        }
        return _field == other._field;
    }

    public override int GetHashCode() => _field;

    // Adding this attribute doesn't seem to help. 
    // I still see the call instruction in the asm regardless (i.e. the method not getting inlined).
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool operator ==(Foo left, Foo right)
    {
        // Test "right" first to allow branch elimination when inlined for null checks (== null) 
        // so it can become a simple test 
        if (right is null)
        {
            // return true/false not the test result https://github.com/dotnet/coreclr/issues/914 
            return (left is null) ? true : false;
        }

        // Changed this to force-inline Equals (rather than doing right.Equals(left);
        if (left == null)
        {
            return false;
        }
        return right._field == left._field;
    }

    public static bool operator !=(Foo left, Foo right)
        => !(left == right);
}
```

I called it in a loop (to try to avoid any JIT tiering issues):
```C#
[Fact]
public static void InliningTest()
{
    var node1 = new Foo(1);
    var node2 = new Foo(1);
    bool res = false;
    for (int i = 0; i < 1_000_000; i++)
    {
        res ^= node1 == node2; // Not getting inlined
        if (res && i > 10_000 && i % 10_000 == 0)
        {
            node1 = new Foo(3);
        }
    }
}
```

Looking at the disassembly, it might even hurt:
https://www.diffchecker.com/s8Keeilh

cc @AndyAyersMS

Reply:
The code above looks like it has a bug, `operator==` calls itself recursively. It's the recursive call that does not get inlined.
```
Inlines into 06000007 X:InliningTest()
  [1 IL=0001 TR=000007 06000001] [below ALWAYS_INLINE size] Foo:.ctor(int):this
    [2 IL=0001 TR=000092 06000491] [below ALWAYS_INLINE size] System.Object:.ctor():this
  [3 IL=0008 TR=000021 06000001] [below ALWAYS_INLINE size] Foo:.ctor(int):this
    [4 IL=0001 TR=000101 06000491] [below ALWAYS_INLINE size] System.Object:.ctor():this
  [5 IL=0023 TR=000044 06000005] [aggressive inline attribute] Foo:op_Equality(ref,ref):bool
    [0 IL=0012 TR=000115 06000005] [FAILED: recursive] Foo:op_Equality(ref,ref):bool
  [6 IL=0051 TR=000083 06000001] [below ALWAYS_INLINE size] Foo:.ctor(int):this
    [7 IL=0001 TR=000158 06000491] [below ALWAYS_INLINE size] System.Object:.ctor():this
```

Reply:
Good catch. I fixed the recursive call, but still see call.

```C#
public static bool operator ==(Foo left, Foo right)
{
    // Test "right" first to allow branch elimination when inlined for null checks (== null) 
    // so it can become a simple test 
    if (right is null)
    {
        // return true/false not the test result https://github.com/dotnet/coreclr/issues/914 
        return (left is null) ? true : false;
    }

    if (left is null)
    {
        return false;
    }
    return right._field == left._field;
}
```

Reply:
What you are doing to look at the jitted codgen? It may be misleading you. The only calls left in the method are to new helpers.
```asm
Inlines into 06000007 X:InliningTest()
  [1 IL=0001 TR=000007 06000001] [below ALWAYS_INLINE size] Foo:.ctor(int):this
    [2 IL=0001 TR=000092 06000491] [below ALWAYS_INLINE size] System.Object:.ctor():this
  [3 IL=0008 TR=000021 06000001] [below ALWAYS_INLINE size] Foo:.ctor(int):this
    [4 IL=0001 TR=000101 06000491] [below ALWAYS_INLINE size] System.Object:.ctor():this
  [5 IL=0023 TR=000044 06000005] [aggressive inline attribute] Foo:op_Equality(ref,ref):bool
  [6 IL=0051 TR=000083 06000001] [below ALWAYS_INLINE size] Foo:.ctor(int):this
    [7 IL=0001 TR=000151 06000491] [below ALWAYS_INLINE size] System.Object:.ctor():this
; Assembly listing for method X:InliningTest()
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 loc0         [V00,T05] (  4,  6   )     ref  ->  rsi         class-hnd
;  V01 loc1         [V01,T08] (  2,  3   )     ref  ->  rdi         class-hnd exact
;  V02 loc2         [V02,T01] (  4, 13   )    bool  ->  rbx
;  V03 loc3         [V03,T00] (  7, 19   )     int  ->  rbp
;  V04 OutArgs      [V04    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V05 tmp1         [V05,T06] (  3,  6   )     ref  ->  rsi         class-hnd exact "NewObj constructor temp"
;  V06 tmp2         [V06,T07] (  3,  6   )     ref  ->  rdi         class-hnd exact "NewObj constructor temp"
;  V07 tmp3         [V07,T02] (  3, 12   )     ref  ->  rsi         class-hnd exact "NewObj constructor temp"
;  V08 tmp4         [V08,T04] (  3,  8   )    bool  ->  rax         "Inline return value spill temp"
;  V09 rat0         [V09,T03] (  3, 12   )     int  ->  rdx         "ReplaceWithLclVar is creating a new local variable"
;
; Lcl frame size = 40

G_M60859_IG01:
       57                   push     rdi
       56                   push     rsi
       55                   push     rbp
       53                   push     rbx
       4883EC28             sub      rsp, 40

G_M60859_IG02:
       48B98030CA9FFD7F0000 mov      rcx, 0x7FFD9FCA3080
       E8793B8C5F           call     CORINFO_HELP_NEWSFAST
       488BF0               mov      rsi, rax
       C7460801000000       mov      dword ptr [rsi+8], 1
       48B98030CA9FFD7F0000 mov      rcx, 0x7FFD9FCA3080
       E8603B8C5F           call     CORINFO_HELP_NEWSFAST
       488BF8               mov      rdi, rax
       C7470801000000       mov      dword ptr [rdi+8], 1
       33DB                 xor      ebx, ebx
       33ED                 xor      ebp, ebp

G_M60859_IG03:
       4885F6               test     rsi, rsi
       7504                 jne      SHORT G_M60859_IG04
       33C0                 xor      eax, eax
       EB0C                 jmp      SHORT G_M60859_IG05

G_M60859_IG04:
       8B4708               mov      eax, dword ptr [rdi+8]
       3B4608               cmp      eax, dword ptr [rsi+8]
       0F94C0               sete     al
       0FB6C0               movzx    rax, al

G_M60859_IG05:
       33C3                 xor      eax, ebx
       0FB6D8               movzx    rbx, al
       85DB                 test     ebx, ebx
       7440                 je       SHORT G_M60859_IG06
       81FD10270000         cmp      ebp, 0x2710
       7E38                 jle      SHORT G_M60859_IG06
       BAAD8BDB68           mov      edx, 0x68DB8BAD
       8BC2                 mov      eax, edx
       F7ED                 imul     edx:eax, ebp
       8BCA                 mov      ecx, edx
       C1E91F               shr      ecx, 31
       C1FA0C               sar      edx, 12
       03CA                 add      ecx, edx
       69C910270000         imul     ecx, ecx, 0x2710
       8BC5                 mov      eax, ebp
       2BC1                 sub      eax, ecx
       7519                 jne      SHORT G_M60859_IG06
       48B98030CA9FFD7F0000 mov      rcx, 0x7FFD9FCA3080
       E8FE3A8C5F           call     CORINFO_HELP_NEWSFAST
       488BF0               mov      rsi, rax
       C7460803000000       mov      dword ptr [rsi+8], 3

G_M60859_IG06:
       FFC5                 inc      ebp
       81FD40420F00         cmp      ebp, 0xF4240
       7C98                 jl       SHORT G_M60859_IG03

G_M60859_IG07:
       4883C428             add      rsp, 40
       5B                   pop      rbx
       5D                   pop      rbp
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret

; Total bytes of code 175, prolog size 8 for method X:InliningTest()
```

Reply:
Based on offline discussion, looks like we need the aggressive inlining attribute for the method to get inlined. Without it, the method won't get inlined. 

However, we can defer annotating this to later once we have some evidence that it will have perf impact and depending on how the type definition/implementation evolves in the coming months (and hence the equality impl). If it remains as simple as a single field comparison, we would add the attribute. But if it starts to get more complex, making it AI could introduce unnecessary code bloat.

Reply:
I created an issue: https://github.com/dotnet/corefx/issues/40186